pulp_ansible/app/galaxy/v1/tasks.py

```diff
@@ -0,0 +1,184 @@
+        pid = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        if pid.returncode != 0:
+            raise Exception(pid.stdout.decode("utf-8"))
+        """
+        clone_url = f"https://github.com/{github_user}/{github_repo}"
+        gitrepo = Repo.clone_from(clone_url, checkout_path, multi_options=["--recurse-submodules"])
+
+        # use the reference to set the version or enumerate it via the util function
+        if github_reference is not None:
+            # cmd = f"git -c advice.detachedHead=false checkout tags/{github_reference}"
+            # pid = subprocess.run(cmd, shell=True, cwd=checkout_path)
+            gitrepo.checkout(github_reference)
+        else:
+            """
+            github_reference = get_role_version(
+                checkout_path=checkout_path,
+                github_user=github_user,
+                github_repo=github_repo,
+                github_reference=github_reference,
+                alternate_role_name=alternate_role_name,
+            )
+            """
+            if gitrepo.tags:
+                github_reference = gitrepo.tags[-1].name
+
+        """
+        # if no tags, default to the commit hash?
+        # WARNING: many roles upstream have -no- versions.
+        if github_reference is None:
+            github_reference = gitrepo.head.commit.hexsha
+        """
+
+        # Upstream stores this, pulp_ansible does not.
+        # github_commit = get_tag_commit_hash(
+        #     clone_url, github_reference, checkout_path=checkout_path
+        # )
+
+        # Upstream stores this as a "release date", pulp_ansible does not.
+        # github_commit_date = get_tag_commit_date(
+        #     clone_url, github_reference, checkout_path=checkout_path
+        # )
+
+        # pulp_ansible does not have role tags
+        """
+        # parse out the role metadata
+        role_meta = get_path_role_meta(checkout_path)
+        role_tags = role_meta.get("galaxy_info", {}).get("galaxy_tags", [])
+        if role_tags is None:
+            role_tags = []
+        """
+
+        # pulp_ansible does not have a model or a view for readme content
+        """
+        ldr = RoleLoader(
+            content_type="role",
+            root=os.path.dirname(checkout_path),
+            rel_path=os.path.basename(checkout_path),
+        )
+        readme = ldr._get_readme()
+        readme_html = markup_utils._render_from_markdown(readme)
+        """
+
+        metadata = {
+            "github_user": github_user,
+            "github_repo": github_repo,
+            "role_name": role_name,
+            "name": role_name,
+            "namespace": github_user,
+            "role_version": github_reference,
+        }
+
+        # pulp_ansible's Roles are actuallly Role Versions
+        # The roles list viewset aggregates the versions together
+        # by using a {namespace}.{name} as the primary key.
+        role = Role(
```

Comment:
I believe it would be good to put the `Role` and `ContentArtifact` into one transaction. `Artifacts` saving can take a long time if Pulp is configured with an object storage that has high latency, so I think you want to make that one outside of the transaction and `save()` it. Pulp's orphan cleanup handles deleting unattached Artifacts so it's safe to make it ahead of time even if the execution stops just after it. Additionally, if it did fail, and the import was attempted again, and it made the same Artifact you could catch the duplicate error from the DB and reuse the existing artifact.

Reply:
The last part about catching the duplicate is a rare case and can be ignored for now I think.